File: rust/conjuredsp-rs/src/buffers.rs
/// Fixed-size circular delay buffer.
///
/// `SIZE` must be a compile-time constant. Store in `static mut` for persistence
/// across callbacks.
///
/// ```ignore
/// static mut DELAYS: [DelayLine<48000>; 2] = [DelayLine::new(); 2];
/// ```
#[derive(Clone, Copy)]
pub struct DelayLine<const SIZE: usize> {
    buf: [f32; SIZE],
    write_pos: usize,
}

impl<const SIZE: usize> DelayLine<SIZE> {
    /// Create a zeroed delay line.
    pub const fn new() -> Self {
        DelayLine {
            buf: [0.0; SIZE],
            write_pos: 0,
        }
    }

    /// Write a sample and advance the write head.
    #[inline]
    pub fn write(&mut self, sample: f32) {
        self.buf[self.write_pos] = sample;
        self.write_pos = (self.write_pos + 1) % SIZE;
    }

    /// Read from the delay line with linear interpolation.
    #[inline]
    pub fn read(&self, delay_samples: f64) -> f32 {
        let read_pos = self.write_pos as f64 - delay_samples;
        let read_pos = if read_pos < 0.0 {
            read_pos + SIZE as f64
        } else {
            read_pos
        };
        let idx = read_pos as usize;
        let frac = (read_pos - idx as f64) as f32;
        let idx0 = idx % SIZE;
        let idx1 = (idx + 1) % SIZE;
        self.buf[idx0] * (1.0 - frac) + self.buf[idx1] * frac
    }

    /// Read with cubic (Hermite) interpolation.
    ///
    /// Better quality than linear for pitch shifting and modulated delays.
    #[inline]
    pub fn read_cubic(&self, delay_samples: f64) -> f32 {
        let read_pos = self.write_pos as f64 - delay_samples;
        let read_pos = if read_pos < 0.0 {
            read_pos + SIZE as f64
        } else {
            read_pos
        };
        let idx = read_pos as usize;
        let frac = (read_pos - idx as f64) as f32;

        let y0 = self.buf[(idx + SIZE - 1) % SIZE];
        let y1 = self.buf[idx % SIZE];
        let y2 = self.buf[(idx + 1) % SIZE];
        let y3 = self.buf[(idx + 2) % SIZE];

        let c0 = y1;
        let c1 = 0.5 * (y2 - y0);
        let c2 = y0 - 2.5 * y1 + 2.0 * y2 - 0.5 * y3;
        let c3 = 0.5 * (y3 - y0) + 1.5 * (y1 - y2);
        c0 + frac * (c1 + frac * (c2 + frac * c3))
    }

    /// Read at integer delay (no interpolation).
    #[inline]
    pub fn tap(&self, delay_samples: usize) -> f32 {
        self.buf[(self.write_pos + SIZE - delay_samples) % SIZE]
    }

    /// Zero the entire buffer and reset write position.
    pub fn clear(&mut self) {
        self.buf = [0.0; SIZE];
        self.write_pos = 0;
    }
}

#[cfg(test)]
mod tests {
    extern crate std;
    use super::*;

    #[test]
    fn test_new_reads_zeros() {
        let dl: DelayLine<64> = DelayLine::new();
        for i in 1..=64 {
            assert_eq!(dl.tap(i), 0.0);
        }
    }

    #[test]
    fn test_write_then_tap1() {
        let mut dl: DelayLine<64> = DelayLine::new();
        dl.write(0.42);
        assert_eq!(dl.tap(1), 0.42);
    }

    #[test]
    fn test_write_n_then_tap_n_wraps() {
        let mut dl: DelayLine<8> = DelayLine::new();
        // Write 8 samples (fills entire buffer), then write 3 more to wrap
        for i in 0..11 {
            dl.write(i as f32);
        }
        // The last sample written was 10.0
        assert_eq!(dl.tap(1), 10.0);
        // 8 samples ago was sample 3.0
        assert_eq!(dl.tap(8), 3.0);
    }

    #[test]
    fn test_read_integer_delay_matches_tap() {
        let mut dl: DelayLine<64> = DelayLine::new();
        for i in 0..20 {
            dl.write(i as f32 * 0.1);
        }
        for delay in 1..=20 {
            let tap_val = dl.tap(delay);
            let read_val = dl.read(delay as f64);
            assert!(
                (tap_val - read_val).abs() < 1e-6,
                "delay={}: tap={} read={}",
                delay,
                tap_val,
                read_val
            );
        }
    }

    #[test]
    fn test_read_fractional_delay_interpolates() {
        let mut dl: DelayLine<64> = DelayLine::new();
        // Write a known sequence: 0, 1, 2, 3, ...
        for i in 0..10 {
            dl.write(i as f32);
        }
        // tap(1) = 9.0, tap(2) = 8.0
        // read(1.5) should interpolate between them: 8.5
        let val = dl.read(1.5);
        assert!(
            (val - 8.5).abs() < 1e-4,
            "fractional read expected 8.5, got {}",
            val
        );
    }

    #[test]
    fn test_read_cubic_integer_delay_matches_tap() {
        let mut dl: DelayLine<64> = DelayLine::new();
        for i in 0..20 {
            dl.write(i as f32 * 0.1);
        }
        for delay in 2..=18 {
            let tap_val = dl.tap(delay);
            let cubic_val = dl.read_cubic(delay as f64);
            assert!(
                (tap_val - cubic_val).abs() < 1e-4,
                "delay={}: tap={} cubic={}",
                delay,
                tap_val,
                cubic_val
            );
        }
    }

    #[test]
    fn test_clear_zeros_everything() {
        let mut dl: DelayLine<16> = DelayLine::new();
        for i in 0..16 {
            dl.write((i + 1) as f32);
        }
        dl.clear();
        for i in 1..=16 {
            assert_eq!(dl.tap(i), 0.0, "tap({}) should be 0 after clear", i);
        }
    }
}

File: rust/conjuredsp/buffers.py
"""Circular buffer and delay line for ConjureDSP presets.

Numpy-backed for efficient memory use and compatibility with the
audio buffer types used throughout ConjureDSP.
"""

import numpy as np


class DelayLine:
    """Pre-allocated circular buffer for delay effects.

    Backed by a numpy float32 array. Supports fractional-sample reads
    via linear or cubic interpolation.

    Example usage in a delay effect::

        _delays = None

        def process(ctx):
            global _delays
            n_ch, frame_count = ctx.inputs.shape
            if _delays is None or len(_delays) != n_ch:
                max_samples = int(0.5 * ctx.sample_rate)  # 500ms max
                _delays = [DelayLine(max_samples) for _ in range(n_ch)]

            delay_samples = ctx.params["time"] * 0.001 * ctx.sample_rate
            feedback = ctx.params["feedback"]
            mix = ctx.params["mix"]

            for ch in range(n_ch):
                dl = _delays[ch]
                row_in = ctx.inputs[ch]
                row_out = ctx.outputs[ch]
                for i in range(frame_count):
                    delayed = dl.read(delay_samples)
                    dl.write(row_in[i] + delayed * feedback)
                    row_out[i] = row_in[i] * (1 - mix) + delayed * mix
    """

    __slots__ = ("_buf", "_size", "_write_pos")

    def __init__(self, max_samples: int) -> None:
        """Allocate a delay line with the given maximum length in samples."""
        self._buf = np.zeros(max_samples, dtype=np.float32)
        self._size = max_samples
        self._write_pos = 0

    @property
    def max_samples(self) -> int:
        """Maximum delay length in samples."""
        return self._size

    def write(self, sample: float) -> None:
        """Write a sample and advance the write head."""
        self._buf[self._write_pos] = sample
        self._write_pos = (self._write_pos + 1) % self._size

    def read(self, delay_samples: float) -> float:
        """Read from the delay line with linear interpolation.

        Args:
            delay_samples: Delay in samples (can be fractional).

        Returns:
            Interpolated sample value.
        """
        read_pos = self._write_pos - delay_samples
        if read_pos < 0.0:
            read_pos += self._size
        idx = int(read_pos)
        frac = read_pos - idx
        idx0 = idx % self._size
        idx1 = (idx + 1) % self._size
        return float(self._buf[idx0] * (1.0 - frac) + self._buf[idx1] * frac)

    def read_cubic(self, delay_samples: float) -> float:
        """Read with cubic (Hermite) interpolation.

        Better quality than linear for pitch shifting and modulated delays.
        Uses 4-point Hermite interpolation.

        Args:
            delay_samples: Delay in samples (can be fractional).

        Returns:
            Interpolated sample value.
        """
        read_pos = self._write_pos - delay_samples
        if read_pos < 0.0:
            read_pos += self._size
        idx = int(read_pos)
        frac = read_pos - idx

        # Four sample points for cubic interpolation
        s = self._size
        buf = self._buf
        y0 = float(buf[(idx - 1) % s])
        y1 = float(buf[idx % s])
        y2 = float(buf[(idx + 1) % s])
        y3 = float(buf[(idx + 2) % s])

        # Hermite interpolation
        c0 = y1
        c1 = 0.5 * (y2 - y0)
        c2 = y0 - 2.5 * y1 + 2.0 * y2 - 0.5 * y3
        c3 = 0.5 * (y3 - y0) + 1.5 * (y1 - y2)
        return c0 + frac * (c1 + frac * (c2 + frac * c3))

    def tap(self, delay_samples: int) -> float:
        """Read at an integer delay (no interpolation).

        Args:
            delay_samples: Delay in whole samples.

        Returns:
            Sample value at the given delay.
        """
        idx = (self._write_pos - delay_samples) % self._size
        return float(self._buf[idx])

    def clear(self) -> None:
        """Zero the entire buffer and reset write position."""
        self._buf.fill(0.0)
        self._write_pos = 0
